src/syscon.rs

```diff
@@ -1,36 +1,469 @@
+        }
+        Ok(())
+    }
 
-    pub fn enable_hace(&mut self) {
-        unsafe {
-            self.scu
-                .scu084()
-                .write(|w| w.scu080clk_stop_ctrl_clear_reg().bits(1 << 13));
+    fn reset_deassert(&mut self, reset_id: u8) -> Result<(), Error> {
+        let mut bit_pos = reset_id;
+        if bit_pos >= ASPEED_RESET_GRP_1_OFFSET + 32 {
+            return Err(Error::InvalidResetId);
+        }
 
-            self.delay.delay_ns(1_000_000);
+        let reg_value: u32 = if bit_pos >= ASPEED_RESET_GRP_1_OFFSET {
+            bit_pos -= ASPEED_RESET_GRP_1_OFFSET;
+            self.scu.scu054().write(|w| unsafe { w.bits(1 << bit_pos) });
+            self.scu.scu054().read().bits()
+        } else {
+            self.scu.scu044().write(|w| unsafe { w.bits(1 << bit_pos) });
+            self.scu.scu044().read().bits()
+        };
 
-            // Release the hace reset
-            self.scu.scu044().write(|w| w.bits(0x10));
+        if reg_value & (1 << bit_pos) != (1 << bit_pos) {
+            return Err(Error::HardwareFailure);
         }
+        Ok(())
     }
-    pub fn enable_rsa_ecc(&mut self) {
-        unsafe {
-            // Enable RSA/ECC clock
-            self.scu
-                .scu094()
-                .write(|w| w.scu090clk_stop_ctrl_clear_reg_set2().bits(1 << 6));
+
+    fn reset_pulse(&mut self, reset_id: u8, duration: Duration) -> Result<(), Error> {
+        let mut result: Result<(), Error>;
+        let bit_pos: u8 = reset_id;
+        if bit_pos >= ASPEED_RESET_GRP_1_OFFSET + 32 {
+            return Err(Error::InvalidResetId);
+        }
+        result = self.reset_assert(reset_id);
+        if result == Ok(()) {
+            self.delay
+                .delay_ns(u32::try_from(duration.as_nanos()).unwrap());
+            result = self.reset_deassert(reset_id);
         }
+        result
+    }
+
+    fn reset_is_asserted(&self, reset_id: u8) -> Result<bool, Error> {
+        let mut bit_pos: u8 = reset_id;
+
+        if bit_pos >= ASPEED_RESET_GRP_1_OFFSET + 32 {
+            return Err(Error::InvalidResetId);
+        }
+
+        let reg_value: u32 = if bit_pos >= ASPEED_RESET_GRP_1_OFFSET + 32 {
+            return Err(Error::InvalidResetId);
+        } else if bit_pos >= ASPEED_RESET_GRP_1_OFFSET {
+            bit_pos -= ASPEED_RESET_GRP_1_OFFSET;
+            self.scu.scu050().read().bits()
+        } else {
+            self.scu.scu040().read().bits()
+        };
+
+        if reg_value & (1 << bit_pos) == (1 << bit_pos) {
+            return Ok(true);
+        }
+        Ok(false)
+    }
+}
+
+#[derive(Clone, PartialEq, Debug)]
+pub struct ClockConfig {
+    pub frequency_hz: u64,
+    pub clk_source_sel: u8,
+}
+
+impl<D: DelayNs> ClockControl for SysCon<D> {
+    type ClockId = ClockId;
+    type ClockConfig = ClockConfig;
+
+    fn enable(&mut self, clock_id: &Self::ClockId) -> Result<(), Self::Error> {
+        match clock_id {
```

Comment:
I noticed that the enable and disable methods use a match statement to handle different ClockId variants. Since all variants are handled in the same way, you can simplify the code by directly calling enable_clock and disable_clock without matching against specific variants. This will make the code more concise and easier to maintain.

```suggestion
impl<D: DelayNs> ClockControl for SysCon<D> {
    type ClockId = ClockId;
    type ClockConfig = ClockConfig;

    fn enable(&mut self, clock_id: &Self::ClockId) -> Result<(), Self::Error> {
        self.enable_clock(*clock_id as u8)
    }

    fn disable(&mut self, clock_id: &Self::ClockId) -> Result<(), Self::Error> {
        self.disable_clock(*clock_id as u8)
    }
}
```